aten/src/ATen/native/Embedding.cpp

```diff
@@ -81,31 +81,52 @@ Tensor embedding_sparse_backward(
   return sparse_type._sparse_coo_tensor_unsafe(index, values, weight_size);
 }
 
-Tensor embedding_dense_backward_cpu(
+Tensor embedding_dense_backward(
     const Tensor & grad_, const Tensor & indices, int64_t num_weights,
     int64_t padding_idx, bool scale_grad_by_freq) {
 
   auto indices_arg = TensorArg(indices, "indices", 2);
   checkScalarType("embedding_backward", indices_arg, kLong);
 
   auto indices_contig = indices.contiguous();
-  auto indices_data = indices_contig.data<int64_t>();
-  int64_t numel = indices.numel();
+  int64_t numel = indices_contig.numel();
+  auto flat_indices = indices_contig.view(-1);
 
   std::unique_ptr<int64_t[]> counts;
   if (scale_grad_by_freq) {
     counts.reset(new int64_t[num_weights]);
     for (int i = 0; i < numel; i++) {
-      counts[indices_data[i]] = 0;
+      counts[flat_indices[i].toCLong()] = 0;
```

Comment:
Is this run for Cuda as well? This would look very inefficient.

For reference, I would have expected embedding_backward for the scaled by inverse frequency case to look something like backward in the following
```
class MyEmbeddingWithScaling(torch.autograd.Function):
    @staticmethod
    def forward(ctx, inp, weight):
        ctx._weight_shape = weight.shape
        ctx.save_for_backward(inp)
        return torch.nn.functional.embedding(inp, weight)
    @staticmethod
    def backward(ctx, grad_out):
        inp, = ctx.saved_tensors
        inp = inp.continuous().view(-1)
        w_num, w_dim = ctx._weight_shape
        counts = weight.new_zeros(w_num)
        counts.index_add_(0, inp, grad_out.new_ones(1).expand_as(inp))
        freq_weight = 1/counts.index_select(0, inp)
        grad_in = grad_out.new_zeros(w_num, w_dim)
        grad_in.index_add_(0, inp, freq_weight[:, None]*grad_out) # replace by fused op
        return None, grad_in
```
but with a fused op in the index_add_.
(I might have done something wrong, but even the above seems to have hardly any performance difference compared to `F.embedding` over the current implementation in my silly benchmarking with 100000x100 embedding and 10000 indices.)

Reply:
Still have to work on embedding, that commit was just the previous commit. It just came along with the rebasing. But I'll keep all this in mind. Thank You.